lib/core/topologies/replset.js

```diff
@@ -1110,19 +1111,36 @@ ReplSet.prototype.selectServer = function(selector, options, callback) {
     readPreference = options.readPreference || ReadPreference.primary;
   }
 
-  const server = this.s.replicaSetState.pickServer(readPreference);
-  if (server == null) {
-    callback(new MongoError('server selection failed'));
-    return;
-  }
+  let lastError;
+  const start = process.hrtime();
+  const _selectServer = () => {
+    if (calculateDurationInMs(start) >= 10000) {
+      if (lastError != null) {
+        callback(lastError, null);
+      } else {
+        callback(new MongoError('Server selection timed out'));
+      }
 
-  if (!(server instanceof Server)) {
-    callback(server, null);
-    return;
-  }
+      return;
+    }
+
+    const server = this.s.replicaSetState.pickServer(readPreference);
+    if (server == null) {
+      setTimeout(_selectServer, 1000);
```

Comment:
Can we make `1000` a constant?

Reply:
I think you made `10000` a constant, which is also great :) can we make this a constant too?

Reply:
done!